execute/outcome.go

```diff
@@ -121,31 +121,45 @@ func (p *Plugin) getMessagesOutcome(
 	observation exectypes.Observation,
 	previousOutcome exectypes.Outcome,
 ) exectypes.Outcome {
-	commitReports := previousOutcome.PendingCommitReports
+	commitReports := make([]exectypes.CommitData, 0)
 	costlyMessagesSet := mapset.NewSet[cciptypes.Bytes32]()
 	for _, msgID := range observation.CostlyMessages {
 		costlyMessagesSet.Add(msgID)
 	}
 
+	// First ensure that all observed messages has hashes and token data.
+	if err := validateHashesExist(observation.Messages, observation.Hashes); err != nil {
+		p.lggr.Errorw("validate hashes exist: %w", err)
+		return exectypes.Outcome{}
+	}
+	if err := validateTokenDataObservations(observation.Messages, observation.TokenData); err != nil {
+		p.lggr.Errorw("validate token data observations: %w", err)
+		return exectypes.Outcome{}
+	}
+
+	reports := observation.CommitReports.Flatten()
 	// add messages to their commitReports.
-	for i, report := range commitReports {
+	for i, report := range reports {
 		report.Messages = nil
 		report.CostlyMessages = nil
 		for j := report.SequenceNumberRange.Start(); j <= report.SequenceNumberRange.End(); j++ {
 			if msg, ok := observation.Messages[report.SourceChain][j]; ok {
+				// Always add the message to the report, even if it is pseudo-deleted.
 				report.Messages = append(report.Messages, msg)
+
+				report.Hashes = append(report.Hashes, observation.Hashes[report.SourceChain][j])
+
 				if costlyMessagesSet.Contains(msg.Header.MessageID) {
 					report.CostlyMessages = append(report.CostlyMessages, msg.Header.MessageID)
 				}
 			}
-
-			if tokenData, ok := observation.TokenData[report.SourceChain][j]; ok {
-				report.MessageTokenData = append(report.MessageTokenData, tokenData)
-			}
+			report.MessageTokenData = append(report.MessageTokenData, observation.TokenData[report.SourceChain][j])
```

Comment:
This is the same deal as message and hash, need an entry for each message. Could it be moved up?

Reply:
Thought I moved it up already. Thanks!